File: bugbounty_tool.py
#!/usr/bin/env python3
"""
Bug Bounty Report Submission Tool
For use with the Immutable Bug Bounty Program
"""

def submit_report(title, description, severity, poc_url=None):
    """Submit a vulnerability report"""
    
    # Validate required fields
    if not title or not description or not severity:
        return "Error: Missing required fields"
    
    # Validate severity level
    valid_severities = ['P1', 'P2', 'P3', 'P4', 'P5']
    if severity not in valid_severities:
        return f"Error: Invalid severity. Must be one of {valid_severities}"
    
    # Format report
    report = {
        'title': title,
        'description': description,
        'severity': severity,
        'poc_url': poc_url,
        'username': get_username(),
        'timestamp': get_timestamp()
    }
    
    # Submit to Bugcrowd API
    try:
        response = requests.post(
            'https://api.bugcrowd.com/vulnerabilities',
            headers={'Content-Type': 'application/json'},
            data=json.dumps(report),
            auth=(get_api_key(), '')
        )
        
        if response.status_code == 201:
            return f"Report submitted successfully! Vulnerability ID: {response.json()['id']}"
        else:
            return f"Error submitting report: {response.status_code} - {response.text}"
            
    except Exception as e:
        return f"Error connecting to Bugcrowd API: {str(e)}"

def get_username():
    """Get researcher's username"""
    return os.environ.get('BUGCROWD_USERNAME', '')

def get_api_key():
    """Get researcher's API key"""
    return os.environ.get('BUGCROWD_API_KEY', '')

def get_timestamp():
    """Get current timestamp"""
    return datetime.datetime.now().isoformat()

def validate_report(title, description, severity, poc_url=None):
    """Validate report before submission"""
    
    # Check length constraints
    if len(title) > 100:
        return "Error: Title exceeds maximum length of 100 characters"
    
    if len(description) > 5000:
        return "Error: Description exceeds maximum length of 5000 characters"
    
    # Check for required fields
    if not title:
        return "Error: Title cannot be empty"
    
    if not description:
        return "Error: Description cannot be empty"
    
    # Validate URL if provided
    if poc_url and not is_valid_url(poc_url):
        return "Error: Provided URL is invalid"
    
    return None

def is_valid_url(url):
    """Validate URL format"""
    try:
        result = urlparse(url)
        return all([result.scheme, result.netloc])
    except:
        return False

def main():
    # Get report details
    title = input("Enter vulnerability title: ")
    description = input("Enter vulnerability description: ")
    severity = input("Enter severity (P1-P5): ")
    poc_url = input("Enter proof-of-concept URL (optional): ")
    
    # Validate before submission
    validation_error = validate_report(title, description, severity, poc_url)
    if validation_error:
        print(validation_error)
        return
    
    # Submit report
    result = submit_report(title, description, severity, poc_url)
    print(result)

if __name__ == "__main__":
    main()

File: bugbounty_tool.js
#!/usr/bin/env node
/**
 * Bug Bounty Report Submission Tool
 * For use with the Immutable Bug Bounty Program
 */

const axios = require('axios');
const readline = require('readline');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Set up environment variables
const BUGCROWD_USERNAME = process.env.BUGCROWD_sourceeye;
const BUGCROWD_API_KEY = process.env.BUGCROWD_API_KEY;

// Validate required environment variables
if (!BUGCROWD_USERNAME || !BUGCROWD_API_KEY) {
  console.error('Error: Missing required environment variables');
  console.log('Please set BUGCROWD_USERNAME and BUGCROWD_API_KEY');
  process.exit(1);
}

// Set up readline interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Function to validate report
function validateReport(title, description, severity, pocUrl) {
  // Check length constraints
  if (title.length > 100) {
    return 'Error: Title exceeds maximum length of 100 characters';
  }
  if (description.length > 5000) {
    return 'Error: Description exceeds maximum length of 5000 characters';
  }
  
  // Check for required fields
  if (!title) return 'Error: Title cannot be empty';
  if (!description) return 'Error: Description cannot be empty';
  
  // Validate URL if provided
  if (pocUrl && !isValidUrl(pocUrl)) {
    return 'Error: Provided URL is invalid';
  }
  
  return null;
}

// Function to validate URL
function isValidUrl(url) {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

// Function to submit report
async function submitReport(title, description, severity, pocUrl) {
  // Format report
  const report = {
    title,
    description,
    severity,
    pocUrl,
    username: BUGCROWD_USERNAME,
    timestamp: new Date().toISOString()
  };
  
  // Submit to Bugcrowd API
  try {
    const response = await axios.post(
      'https://api.bugcrowd.com/vulnerabilities',
      report,
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${BUGCROWD_API_KEY}`
        }
      }
    );
    
    return `Report submitted successfully! Vulnerability ID: ${response.data.id}`;
    
  } catch (error) {
    return `Error submitting report: ${error.response?.status} - ${error.response?.data}`;
  }
}

// Main function
async function main() {
  // Get report details
  const title = await new Promise(resolve => 
    rl.question('Enter vulnerability title: ', resolve)
  );
  
  const description = await new Promise(resolve => 
    rl.question('Enter vulnerability description: ', resolve)
  );
  
  const severity = await new Promise(resolve => 
    rl.question('Enter severity (P1-P5): ', resolve)
  );
  
  const pocUrl = await new Promise(resolve => 
    rl.question('Enter proof-of-concept URL (optional): ', resolve)
  );
  
  // Validate before submission
  const validationError = validateReport(title, description, severity, pocUrl);
  if (validationError) {
    console.error(validationError);
    return;
  }
  
  // Submit report
  const result = await submitReport(title, description, severity, pocUrl);
  console.log(result);
  
  // Close readline interface
  rl.close();
}

// Run main function
main().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});
